File: LayoutDotSyntaxExample/JGDynamicSizeLabel.m
//
//  JGDynamicSizeLabel.m
//  LayoutDotSyntaxExample
//
//  Created by Jaden Geller on 3/9/14.
//  Copyright (c) 2014 Jaden Geller. All rights reserved.
//

#import "JGDynamicSizeLabel.h"
#import "JGLayoutParameter.h"

@interface JGDynamicSizeLabel ()
{
    UIFont *_font;
}

@property (nonatomic) JGLayoutParameter *parameter;

@end

@implementation JGDynamicSizeLabel

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        // Initialization code
    }
    return self;
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect
{
    // Drawing code
}
*/

-(void)setFont:(UIFont *)font{
    _font = font;
    
   // >> UPDATE FONT
}

-(UIFont*)font{
    if (!_font) {
        _font = [super font];
    }
    return _font;
}

-(void)setParameter:(JGLayoutParameter *)parameter{
    if (_parameter) [self clean_JGDynamicSizeLabel];
    _parameter = parameter;
    
    [[parameter.object layer] addObserver:self forKeyPath:@"bounds" options:NSKeyValueObservingOptionNew context:NULL];
}
         
-(void)setFontSize:(id)fontSizeMultiplier{
    if ([fontSizeMultiplier isKindOfClass:[NSNumber class]]) {
        [super setFont:[self.font fontWithSize:self.font.pointSize * [fontSizeMultiplier doubleValue]]];
    }
    else if ([fontSizeMultiplier isKindOfClass:[JGLayoutParameter class]]){
        if ([(JGLayoutParameter*)fontSizeMultiplier attribute] == NSLayoutAttributeWidth || [(JGLayoutParameter*)fontSizeMultiplier attribute] == NSLayoutAttributeHeight) {
            self.parameter = (JGLayoutParameter*)fontSizeMultiplier;
        }
        else [JGDynamicSizeLabel exception_JGDynamicSizeLabel];
    }
    else if (!fontSizeMultiplier) [self clean_JGDynamicSizeLabel];
    else [JGDynamicSizeLabel exception_JGDynamicSizeLabel];
    
    _fontSize = fontSizeMultiplier;
}

-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context{
    if ([object isEqual:[self.parameter.object layer]] && [keyPath isEqualToString:@"bounds"]) {
        CGFloat size = 0;
        if (self.parameter.attribute == NSLayoutAttributeWidth) {
            size = [self.parameter.object bounds].size.width;
        }
        else if (self.parameter.attribute == NSLayoutAttributeHeight){
            size = [self.parameter.object bounds].size.height;
        }
        else{
            [NSException raise:@"Bad font size constraint." format:@"Font size constraint must be either a NSNumber or a JGLayoutParameter."];
        }
        
        [super setFont:[self.font fontWithSize:size * self.parameter.multiplier + self.parameter.constant]];
    }
}

+(void)exception_JGDynamicSizeLabel{
    [NSException raise:@"Bad font size constraint." format:@"Font size constraint must be either a NSNumber or a JGLayoutParameter."];
}

-(void)clean_JGDynamicSizeLabel{
    [[_parameter.object layer] removeObserver:self forKeyPath:@"bounds.size"];
    [super setFont:self.font];
}

-(void)dealloc{
    [self clean_JGDynamicSizeLabel];
}

@end

File: JGLayoutDotSyntax/JGDynamicSizeLabel.swift
//
//  JGDynamicSizeLabel.swift
//  JGLayoutDotSyntax
//
//  Objective-C code Copyright (c) 2014 Jaden Geller. All rights reserved.
//  Swift adaptation Copyright (c) 2014 Nicolas Gomollon. All rights reserved.
//

import Foundation
import UIKit
import QuartzCore

class JGDynamicSizeLabel: UILabel {
	
	/** The font size of the label. Used as a setter, this creates and adds a "constraint" based upon the specification of the assigned JGLayoutParameter. The label's original font size is ignored, and replaced by this constraint. Can only be used with the width and the height attribute. Used as a getter, will return any associated constraint. If none exists, will return nil. Note that, in order to add a font size constraint, the label must be the sender.
 
 @code
 
 label.fontSize = view.layoutHeight // sets viewA's baseline to match that of viewB
 
 @endcode
 
 @see https://github.com/JadenGeller/JGLayoutDotSyntax for more information.
 
	*/
	var fontSize: AnyObject! {
	get {
		return parameter
	}
	set {
		if newValue == nil {
			parameter = nil
		} else if let fontSizeMultiplier = newValue as? JGLayoutParameter {
			if (fontSizeMultiplier.attribute == NSLayoutAttribute.Width) || (fontSizeMultiplier.attribute == NSLayoutAttribute.Height) {
				parameter = fontSizeMultiplier
			}
		} else if let fontSizeMultiplier = newValue as? Double {
			font = font.fontWithSize(font.pointSize * CGFloat(fontSizeMultiplier))
		}
	}
	}
	
	var parameter: JGLayoutParameter! {
	willSet {
		if let p = parameter {
			if var view = p.object as? UIView {
				view.layer.removeObserver(self, forKeyPath:"bounds")
			}
		}
	}
	didSet {
		if let p = parameter {
			if var view = p.object as? UIView {
				view.layer.addObserver(self, forKeyPath: "bounds", options: .New, context: nil)
			}
		}
	}
	}
	
	
	convenience override init() {
		self.init(frame: CGRectZero)
	}
	
	override init(frame: CGRect) {
		super.init(frame: frame)
	}
	
	required init(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
	}
	
	override func observeValueForKeyPath(keyPath: String!, ofObject object: AnyObject!, change: [NSObject : AnyObject]!, context: UnsafeMutablePointer<Void>) {
		if let p = parameter {
			var view = p.object as? UIView
			var viewObject = object as? CALayer
			if (view != nil) && (viewObject == view!.layer) && (keyPath == "bounds") {
				var size: CGFloat = 0.0
				switch p.attribute {
				case .Width:
					size = view!.bounds.size.width
				case .Height:
					size = view!.bounds.size.height
				default:
					break
				}
				font = font.fontWithSize(size * CGFloat(p.multiplier) + CGFloat(p.constant))
			}
		}
	}
	
}
